### PR TITLE
Development task48 priyanka

### PR DESCRIPTION
**Pull request by PriyankaChaganti**
https://
screenshot 9

github.com/PriyankaChaganti/traffic_sign_svm/issues/48

### CODE REVIEW

**Comment by PriyankaChaganti** [comment]
screenshot 10

File: robust_svm/main.py [hunk 1]
```diff
@@ -1,14 +1,59 @@
...
+multi_svm_classifier = MultiClassClassifier(training_data, epochs, svm_params)
+pickle.dump(multi_svm_classifier, open(multi_svm_classifier_pickle_path, "wb"))
+# Comment the above two lines and uncomment the following line to avoid re-building
+# classifier
+# multi_svm_classifier = pickle.load(open(multi_svm_classifier_pickle_path, "rb"))
 
 
+###########################################################################################
+# Test Classifier
+###########################################################################################
```
Comment (sahithi403) [inline robust_svm/main.py:59]: Where is the testing code?

File: robust_svm/svm_utils.py [hunk 2]
```diff
@@ -12,24 +11,64 @@ def check_image_class(row, image_class):
...
-    """
-    
-    :param multi_class_classifier: 
-    :param test_data: 
-    :return: 
-    """
+
+
+def determine_image_class(multiple_labels):
+    labels_sum = sum(multiple_labels.values())
```
Comment (sahithi403) [inline robust_svm/svm_utils.py:26]: Add docstring

File: robust_svm/svm_utils.py [hunk 3]
```diff
@@ -12,24 +11,64 @@ def check_image_class(row, image_class):
...
+    if labels_sum == 1:
+        for class_id, label in multiple_labels.items():
+            if label == 1:
+                return class_id
+
+    return None
+
+
+def test_classifier(multi_class_classifier, test_data):
+    accuracy_results = {}
```
Comment (sahithi403) [inline robust_svm/svm_utils.py:43]: Add docstring

File: robust_svm/main.py [hunk 4]
```diff
@@ -51,9 +51,15 @@
 pickle.dump(multi_svm_classifier, open(multi_svm_classifier_pickle_path, "wb"))
 # Comment the above two lines and uncomment the following line to avoid re-building
 # classifier
-# multi_svm_classifier = pickle.load(open(multi_svm_classifier_pickle_path, "rb"))
+# multi_svm_classifier = pickle.
```
Comment (sahithi403) [inline robust_svm/main.py]: Why did you change this?

File: robust_svm/main.py [hunk 5]
```diff
@@ -51,9 +51,15 @@
...
+# load(open(multi_svm_classifier_pickle_path, "rb"))
 
 
 ###########################################################################################
 # Test Classifier
 ###########################################################################################
+multi_svm_classifier_pickle_path_test = join(settings.dumps_folder,'multi_svm_classifier_test.p')
```
Comment (sahithi403) [inline robust_svm/main.py]: Why are you loading from the pickle when you are not using 'multi_svm_classifier_pickle_path_test' anywhere?. 

The classifier is in the variable 'multi_svm_classifier ' on line 50

File: robust_svm/main.py [hunk 6]
```diff
@@ -51,9 +51,15 @@
...
+# Building classifier using test_data
+multi_svm_classifier_test = MultiClassClassifier(test_data,epochs,svm_params)
```
Comment (sahithi403) [inline robust_svm/main.py]: You build a classifier using training data. Why would you build a classifier using test data?


You are supposed to test the classifier using test data.

File: robust_svm/svm_utils.py [hunk 7]
```diff
@@ -18,6 +18,11 @@ def check_image_class(row, image_class):
 
 
 def determine_image_class(multiple_labels):
+    """
+    The function determines if the given image is classified to a particular class.
+    :param multiple_labels:Results Dictionary from test_classifier.
```
Comment (sahithi403) [inline robust_svm/svm_utils.py]: The parameter multiple_labels is an output of get_all_svm_labels. How did you not know that?

File: robust_svm/classifier.py [hunk 8]
```diff
@@ -129,21 +127,29 @@ def get_svm_label(self, classifier_id, row):
         return label
 
     def get_all_classifier_labels(self,row):
+        """
+        The function gets the lebel of the row for every classifier2 and returns the label as a dictionary.
```
Comment (sahithi403) [inline robust_svm/classifier.py]: Fix the spellings

File: robust_svm/classifier.py [hunk 9]
```diff
@@ -129,21 +126,30 @@ def get_svm_label(self, classifier_id, row):
...
+    training_data_pickle_path = join(settings.dumps_folder, "training_data.p")
+    multi_svm_classifier_pickle_path = join(settings.dumps_folder, 'multi_svm_classifier.p')
     feature_path = hog3_path
     data_set_list = ['00013', '00015']
-    training_data = make_dataset(data_path, feature_path, data_set_list)
-    kernel_type = "linear"
-    ml = MultiClassClassifier(training_data,1,{"r0":1,"c":1,kernel_type:"linear"})
+    training_data = pickle.load(open(training_data_pickle_path, "rb"))
+    #svmclassifier = pickle.load(open(multi_svm_classifier_pickle_path, "rb"))
+    svmclassifier = MultiClassClassifier(training_data,1,{'r0':1,'C':1,"kernel_type":'linear'})
```
Comment (sahithi403) [inline robust_svm/classifier.py]: Comment out the line with instantiation and uncomment the classifier pickle load

File: robust_svm/image_processing_utils.py [hunk 10]
```diff
@@ -208,6 +208,22 @@ def get_neighboring_points(x, y, x_lim, y_lim):
 
     return neighbors
 
+def image_to_array(image):
+    image = cv2.resize(image, (35,35) )
```
Comment (sahithi403) [inline robust_svm/image_processing_utils.py:217]: add docstring

File: robust_svm/image_processing_utils.py [hunk 11]
```diff
@@ -208,6 +208,22 @@ def get_neighboring_points(x, y, x_lim, y_lim):
...
+    image_array = image.ravel()
+    binary_array = image_array > 0
+    binary_array = binary_array.astype(int)
+    return binary_array
+
+def image_to_feature_vector(dataset_path, annotation):
+    cropped_image = get_cropped_image(dataset_path, annotation)
```
Comment (sahithi403) [inline robust_svm/image_processing_utils.py:230]: add docstring

File: robust_svm/settings.py [hunk 12]
```diff
@@ -2,24 +2,29 @@
...
     assert isdir(temp_folder) is True
-    assert isdir(join(training_data_folder,images_path)) is True
-    assert isdir(join(test_data_folder,images_path)) is True
-    assert isdir(join(training_data_folder,hog3_path))is True
-    assert isdir(join(test_data_folder,hog3_path)) is True
-
+    assert isdir(join(training_data_folder, images_path)) is True
+    assert isdir(join(test_data_folder, images_path)) is True
+    assert isdir(join(training_data_folder, hog3_path))is True
+    assert isdir(join(test_data_folder, hog3_path)) is True
```
Comment (sahithi403) [inline robust_svm/settings.py:30]: add asserts for hog_2 in training and test
add asserts for hog_1 in training and test
add asserts for custom_features in training and test

**Review by sahithi403 — APPROVED** [review]
